awpy/analytics/map_control.py

```diff
@@ -0,0 +1,382 @@
+"""Functions for calculating map control values and metrics.
+
+    Example notebook:
+
+    https://github.com/pnxenopoulos/awpy/blob/main/examples/05_Map_Control_Calculations_And_Visualizations.ipynb
+"""
+
+from collections import defaultdict
+
+import numpy as np
+
+from awpy.analytics.nav import area_distance, calculate_tile_area, find_closest_area
+from awpy.data import NAV, NAV_GRAPHS
+from awpy.types import FrameMapControl, GameFrame, GameRound, TeamFrameInfo
+
+
+def approx_neighbors(
+    map_name: str,
+    source_tile_id: int,
+    n_neighbors: int = 5,
+) -> list[tuple[float, int]]:
+    """Approximates neighbors for isolated tiles by finding n closest tiles.
+
+    Args:
+        map_name (str): Map for source_tile_id
+        source_tile_id (int): Area ID as an integer
+        n_neighbors (int): Number of closest tiles/approximated neighbors wanted
+    Returns:
+        List of tuples representing n closest tiles.
+            Tuple[0] -> Distance between source_tile_id and represented tile
+            Tuple[1] -> Area ID of represented tile
+    Raises:
+        ValueError: If map_name is not in awpy.data.NAV
+                    If source_tile_id is not in awpy.data.NAV[map_name]
+                    If the length of point is not 3
+    """
+    if map_name not in NAV:
+        msg = "Map not found."
+        raise ValueError(msg)
+    if source_tile_id not in NAV[map_name].keys():
+        msg = "Area ID not found."
+        raise ValueError(msg)
+    if n_neighbors < 0:
+        msg = "n_neighbors must be >= 0"
+        raise ValueError(msg)
+
+    current_map_info = NAV[map_name]
+    distance_arr: list[tuple[float, int]] = []
+
+    for tile in current_map_info:
+        if tile != source_tile_id:
+            current_distance = area_distance(
+                map_name, tile, source_tile_id, dist_type="euclidean"
+            )["distance"]
+            distance_arr.append((current_distance, tile))
+    distance_arr.sort()
+
+    return distance_arr[:n_neighbors]
+
+
+def bfs_helper(
+    map_name: str,
+    current_tiles: list[int],
+    neighbor_info: dict[int, set[int]],
+) -> dict:
+    """Helper function to run bfs from each tile in current_tiles.
+
+    Args:
+        map_name (str): Map for current_tiles
+        current_tiles (list): List of tiles to use as start tile for bfs iteration(s)
+        neighbor_info (dict): Dictionary mapping tile to its navigable neighbors
+
+    Returns:
+        Dictionary mapping tile id -> list of map control values
+
+    Raises:
+        ValueError: If map_name is not in awpy.data.NAV
+    """
+    if map_name not in NAV:
+        msg = "Map not found."
+        raise ValueError(msg)
+
+    current_dict = defaultdict(list)
+    start_val = 10
+    step_size = 1
```

Comment:
Yeah that makes sense to make an optional parameter with default value 10. And yeah I think the stepsize is unnecessary now; previoulsy I thought step_size could be turned in a optional parameter to modify how map control value is allocated based on tile distance to original tile. Probably confusing but yeah don't think step_size is needed anymore especially since max_depth will be a optional parameter now